tools/testrunner/testrunner.go

```diff
@@ -0,0 +1,348 @@
+		if len(n.children) > 0 {
+			continue
+		}
+		leafFailures = append(leafFailures, path)
+	}
+
+	return leafFailures
+}
+
+type runner struct {
+	gotestsumExecutable string
+	junitOutputPath     string
+	attempts            []*attempt
+	retries             int
+}
+
+func newRunner(gotestsumExecutable string) *runner {
+	return &runner{
+		gotestsumExecutable: gotestsumExecutable,
+		attempts:            make([]*attempt, 0),
+	}
+}
+
+func (r *runner) sanitizeAndParseArgs(args []string) ([]string, error) {
+	var sanitizedArgs []string
+	type action struct {
+		f   func(string) error
+		err string
+	}
+	var next *action
+	for i, arg := range args {
+		if next != nil {
+			if err := next.f(arg); err != nil {
+				return nil, err
+			}
+			next = nil
+			continue
+		} else if arg == "-retries" {
+			next = &action{
+				f: func(arg string) error {
+					var err error
+					r.retries, err = strconv.Atoi(arg)
+					return err
+				},
+				err: "got -retries flag with no value",
+			}
+			continue
+		} else if strings.HasPrefix(arg, "-retries=") {
+			var err error
+			r.retries, err = strconv.Atoi(arg[len("-retries="):])
+			if err != nil {
+				return nil, err
+			}
+			continue
+		} else if arg == "--junitfile" {
+			// --junitfile is used by gotestsum
+			next = &action{
+				f: func(arg string) error {
+					r.junitOutputPath = arg
+					return nil
+				},
+				err: "got --junitfile flag with no value",
+			}
+			continue
+		} else if strings.HasPrefix(arg, "--junitfile=") {
+			// --junitfile is used by gotestsum
+			r.junitOutputPath = arg[len("--junitfile="):]
+			continue
+		} else if arg == "--" {
+			// Forward all arguments from -- on.
+			sanitizedArgs = append(sanitizedArgs, args[i:]...)
+			break
+		}
+		sanitizedArgs = append(sanitizedArgs, arg)
+	}
+	if next != nil {
+		return nil, fmt.Errorf("incomplete command line arguments: %s", next.err)
+	}
+	if r.junitOutputPath == "" {
+		return nil, fmt.Errorf("missing required argument --junitfile")
+	}
+	return sanitizedArgs, nil
+}
+
+func (r *runner) newAttempt() *attempt {
+	f := filepath.Join(os.TempDir(), fmt.Sprintf("temporalio-temporal-%s-junit.xml", uuid.NewString()))
+	a := &attempt{
+		runner:       r,
+		number:       len(r.attempts) + 1,
+		junitXmlPath: f,
+	}
+	r.attempts = append(r.attempts, a)
+	return a
+}
+
+func (r *runner) combineAttempts() junit.Testsuites {
+	var combined junit.Testsuites
+
+	for i, attempt := range r.attempts {
+		combined.Tests += attempt.suites.Tests
+		combined.Errors += attempt.suites.Errors
+		combined.Failures += attempt.suites.Failures
+		combined.Skipped += attempt.suites.Skipped
+		combined.Disabled += attempt.suites.Disabled
+		combined.Time += attempt.suites.Time
+
+		if i == 0 {
+			combined.XMLName = attempt.suites.XMLName
+			combined.Name = attempt.suites.Name
+			combined.Suites = attempt.suites.Suites
+			continue
+		}
+
+		for _, suite := range attempt.suites.Suites {
+			cpy := suite
+			cpy.Name += fmt.Sprintf(" (retry %d)", i)
```

Comment:
will it be (retry 0)?

Reply:
I think I would rather skip it for the first retry

Reply:
A few lines up we have an `if i == 0 { ... continue }`